### ISSUE TITLE
About pointed box to produce its caption

### ISSUE THREAD

**Issue by alice-cool** [soloist97/densecap-pytorch#1]
Dear scholar, 
   I am glad to read your code about densecaptioning. And I want to know if the code can realize that I pass the designated boxes to the net, and the net return the specific boxes' captions?

**Comment by alice-cool** [comment]
Dear scholar,
   IF the code can produce pointed box's caption. you can also pop your code's bibtex reference.

**Comment by soloist97** [comment]
Yes, theoretically, the code can achieve this by passing your designated boxes following the same procedure defined in the pytorch GeneralizedRCNN forward function. 

I copy the code from pytorch and present it here. Following their steps, and change functions like ` self.roi_heads` to `net.roi_heads`. 

```python
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
"""
Implements the Generalized R-CNN framework
"""

from collections import OrderedDict
import torch
from torch import nn


class GeneralizedRCNN(nn.Module):
    """
    Main class for Generalized R-CNN.

    Arguments:
        backbone (nn.Module):
        rpn (nn.Module):
        heads (nn.Module): takes the features + the proposals from the RPN and computes
            detections / masks from it.
        transform (nn.Module): performs the data transformation from the inputs to feed into
            the model
    """

    def __init__(self, backbone, rpn, roi_heads, transform):
        super(GeneralizedRCNN, self).__init__()
        self.transform = transform
        self.backbone = backbone
        self.rpn = rpn
        self.roi_heads = roi_heads

    def forward(self, images, targets=None):
        """
        Arguments:
            images (list[Tensor]): images to be processed
            targets (list[Dict[Tensor]]): ground-truth boxes present in the image (optional)

        Returns:
            result (list[BoxList] or dict[Tensor]): the output from the model.
                During training, it returns a dict[Tensor] which contains the losses.
                During testing, it returns list[BoxList] contains additional fields
                like `scores`, `labels` and `mask` (for Mask R-CNN models).

        """
        if self.training and targets is None:
            raise ValueError("In training mode, targets should be passed")
        original_image_sizes = [img.shape[-2:] for img in images]  # images (N, C, H, W)
        images, targets = self.transform(images, targets)
        features = self.backbone(images.tensors)  # using FPN to obtain a set of tensors, i.e. OrderedDict(Tensor)
        if isinstance(features, torch.Tensor):
            features = OrderedDict([(0, features)])
        proposals, proposal_losses = self.rpn(images, features, targets)  # no need to use rpn, change them with your designated boxes
        # proposals list(Tensor) [(valid_proposals_num, 4), ...]
        # proposal_losses dict  {"loss_objectness":...,"loss_rpn_box_reg":...}
        
        detections, detector_losses = self.roi_heads(features, proposals, images.image_sizes, targets)
        detections = self.transform.postprocess(detections, images.image_sizes, original_image_sizes)

        losses = {}
        losses.update(detector_losses)
        losses.update(proposal_losses)

        if self.training:
            return losses

        return detections
```

https://github.com/soloist97/densecap-pytorch/blob/bc81d9816ff8d4e45613846ad2acdf789acde37b/model/roi_heads.py#L281-L288

I hope this could help:)

**Comment by alice-cool** [comment]
Thanks for your timely reply. Thanks



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2021年8月10日(星期二) 凌晨0:16
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [soloist97/densecap-pytorch] About pointed box to produce its caption (#1)




 
Yes, theoretically, the code can achieve this by passing your designated boxes following the same procedure defined in the pytorch GeneralizedRCNN forward function.
 
I copy the code from pytorch and present it here. Following their steps, and change functions like  self.roi_heads to net.roi_heads.
 # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved. """ Implements the Generalized R-CNN framework """ from collections import OrderedDict import torch from torch import nn class GeneralizedRCNN(nn.Module):     """     Main class for Generalized R-CNN.      Arguments:         backbone (nn.Module):         rpn (nn.Module):         heads (nn.Module): takes the features + the proposals from the RPN and computes             detections / masks from it.         transform (nn.Module): performs the data transformation from the inputs to feed into             the model     """     def __init__(self, backbone, rpn, roi_heads, transform):         super(GeneralizedRCNN, self).__init__()         self.transform = transform         self.backbone = backbone         self.rpn = rpn         self.roi_heads = roi_heads     def forward(self, images, targets=None):         """         Arguments:             images (list[Tensor]): images to be processed             targets (list[Dict[Tensor]]): ground-truth boxes present in the image (optional)          Returns:             result (list[BoxList] or dict[Tensor]): the output from the model.                 During training, it returns a dict[Tensor] which contains the losses.                 During testing, it returns list[BoxList] contains additional fields                 like `scores`, `labels` and `mask` (for Mask R-CNN models).          """         if self.training and targets is None:             raise ValueError("In training mode, targets should be passed")         original_image_sizes = [img.shape[-2:] for img in images]  # images (N, C, H, W)         images, targets = self.transform(images, targets)         features = self.backbone(images.tensors)  # using FPN to obtain a set of tensors, i.e. OrderedDict(Tensor)         if isinstance(features, torch.Tensor):             features = OrderedDict([(0, features)])         proposals, proposal_losses = self.rpn(images, features, targets)  # no need to use rpn, change them with your designated boxes         # proposals list(Tensor) [(valid_proposals_num, 4), ...]         # proposal_losses dict  {"loss_objectness":...,"loss_rpn_box_reg":...}                  detections, detector_losses = self.roi_heads(features, proposals, images.image_sizes, targets)         detections = self.transform.postprocess(detections, images.image_sizes, original_image_sizes)         losses = {}         losses.update(detector_losses)         losses.update(proposal_losses)         if self.training:             return losses         return detections
 
https://github.com/soloist97/densecap-pytorch/blob/bc81d9816ff8d4e45613846ad2acdf789acde37b/model/roi_heads.py#L281-L288
 
I hope this could help:)
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
